File: LeetCode/Problems/141.py
# NeetCode - Linked List
# 141. Linked List Cycle
# Difficulty : Easy
# Algorithm : Linked List
# Time complexity : O(N), Space complexity : O(1)
# Runtime : 56 ms (90.60%), Memory : 20.30 MB (76.27%)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                return True
        return False

File: LeetCode/Problems/141.java
/*
141. Linked List Cycle
Difficulty : Easy
Algorithm : Linked List, Two Pointers
Runtime : 0 ms (100%), Memory : 43.77 MB (98.31%)
*/

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if(head == null) { return false; }
        ListNode slow = head;
        ListNode fast = head.next;

        while(fast != null) {
            if(slow == fast) { return true; }

            if(fast.next != null) {
                slow = slow.next;
                fast = fast.next.next;
            } else { break; }
        }
        return false;
    }
}
